Add support for shields.io-based badges

algernon opened the pull request:
Adds a new `/{username}/{repo}/badges` family of routes, which redirect to various shields.io badges. The goal is to not reimplement badge generation, and delegate it to shields.io (or a similar service), which are already used by many. This way, we get all the goodies that come with it: different styles, colors, logos, you name it.

So these routes are just thin wrappers around shields.io that make it easier to display the information we want. The URL is configurable via `app.ini`, and is templatable, allowing to use alternative badge generator services with slightly different URL patterns.

Fixes #5633 and #23688.

Work sponsored by Codeberg e.V.

---

This is an alternative solution for #23688 (and #5633), doing things in a very different way than #28102.

## Things left to do

- [x] Make the feature toggleable in `app.ini`
- [x] Make the generator URL templateable
- [x] Error handling:
  - Internal errors return 500
  - Simply not finding the appropriate info (because no releases, or repo units being disabled) redirects to a badge with "not found" in the text section, and a crimson background.
- [x] Tests
  - [x] Positive tests (things work)
  - [x] Negative tests (things do not work)
- [x] Documentation

## Sample screenshot

![Screenshot from 2023-12-22 11-43-16](https://github.com/go-gitea/gitea/assets/17243/88ddbc5c-37a1-42c7-bce5-2ff9242e7ec4)

lng2020 commented:
I'm neutral to a more general badge implementation. If this PR is accepted, I will close my PR.

6543 commented:
It should be able to be at lest deactivatable as awhole, because of dataprotection roules

algernon commented:
> It should be able to be at lest deactivatable as awhole, because of dataprotection roules

Can do, no problem.

delvh commented:
Or even better: We let the user configure the shields.io URL, and if it is blank, it is deactivated.

delvh commented:
But yeah, sounds like a good idea overall.

algernon commented:
> Or even better: We let the user configure the badge generation URL (default: https://shields.io), and if it is blank, it is deactivated.

The URL is already configurable in this PR, defaulting to `img.shields.io`. I think I'd prefer an explicit `enable=false` for disabling.

jolheiser commented:
TIL you can host your own shields.io :grin: 

I wonder if it would make sense to have a config URL with placeholders for `{label}` `{text}` and `{color}`?
In practice I'm not sure if that's premature customization or not.

In this case the default setting could be [`https://shields.io/badge/{label}-{text}-{color}`](https://shields.io/badge/label-text-purple) but would allow someone to mix it however they wanted to. 

I just looked up an alternative as an example, [`https://badgen.net/static/{label}/{text}/{color}`](https://badgen.net/static/label/text/purple)

algernon commented:
> TIL you can host your own shields.io 😁
> 
> I wonder if it would make sense to have a config URL with placeholders for `{label}` `{text}` and `{color}`? In practice I'm not sure if that's premature customization or not.

Yeah, that should be doable. There's existing code that does templated URLs (for `ROOT_URL`, for example), I suppose I can use a similar mechanism for this URL too.

algernon commented:
I'm pondering about how to make the badges easy to use in the README... As things stand, you need a full URL, like `https://my-forge.example.com/user/repo/badges/release.svg`. I can't just adjust the renderer to treat `badges/release.svg`-like relative URLs specially, because if a repo *does* have a `badges/release.svg` file, that would not be possible to reference then.

On the other hand, a large part of the badges use case is when they're displayed *elsewhere*. In that case, we need the full URL anyway. So I guess that'll be the way.

algernon commented:
Cleaned up the code, made it possible to disable the feature, and change the badge generator URL template, added a bunch of tests (both positive and negative ones), and an example to `app.example.ini`, and some docs.

lng2020 commented on models/actions/run.go:
```diff
@@ -308,6 +308,17 @@ func InsertRun(ctx context.Context, run *ActionRun, jobs []*jobparser.SingleWork
 	return commiter.Commit()
 }
 
+func GetLatestRunForBranchAndWorkflow(ctx context.Context, repoID int64, branch, workflowFile string) (*ActionRun, error) {
+	var run ActionRun
+	has, err := db.GetEngine(ctx).Where("repo_id=?", repoID).And("ref=?", branch).And("workflow_id=?", workflowFile).Desc("id").Get(&run)
+	if err != nil {
+		return nil, err
+	} else if !has {
+		return nil, util.NewNotExistErrorf("run with repo_id %d, ref %s, workflow_id %s", repoID, branch, workflowFile)
+	}
+	return &run, nil
+}
+
```
But the `schedule` run doesn't have a branch ref?

algernon replied:
The function's purpose is - as the name implies - get a Run for a given branch and workflow. It is not concerned with runs that don't have a branch.

lng2020 replied:
But in your route, `workflows/{workflow_name}/status.svg` uses this underlying function `GetLatestRunForBranchAndWorkflow` with the default branch. If a workflow file that runs can only be triggered by schedule, then no badge will be returned. But IMO, `workflows/{workflow_name}/status.svg` should return the last run status.

algernon replied:
Mmm. That certainly wasn't the use-case for the route I had in mind, but I see your point. Will see what I can do. Perhaps it'd check if a workflow exists for the default branch, and if not, select the latest scheduled one. I suppose that'd satisfy both needs.

Alternatively - and I think I'd like that approach better - I can change the routes:

- `workflows/{workflow_name}/status.svg` would check the main branch.
- `workflows/{workflow_name}/branch/{branch_name}/status.svg` would check a given branch.
- `workflows/{workflow_name}/scheduled/status.svg` would check the latest scheduled run.

What do you think?

denyskon replied:
A scheduled run does have a branch ref if I understand it correctly - it is just sometimes empty due to a bug which should be fixed with #28640

lng2020 replied:
Now scheduled run has a default branch ref, I think the router can be designed to be compatible with GitHub.
https://docs.github.com/en/actions/monitoring-and-troubleshooting-workflows/adding-a-workflow-status-badge
workflows/{workflow_name}/badge.svg with `event` and `branch` para.

algernon replied:
Can do, I'll update the PR later today.

algernon replied:
Hrm. I *can* make it compatible with GitHub, even the same URI, just different host. But I kinda like how every badge lives under `/badges/`. I can make both work, though. I suppose I'll do that, and let others decide which one to keep (or to keep both).

wxiaoguang commented:
TBH, I have strong preference to #28102 if the concerns there could be resolved.

algernon commented:
Force-pushed an update, changing the route of the action badges: it's now `/badges/workflows/{workflow_file}/badge.svg` with optional `branch` and `event` query parameters. Additionally, for closer GitHub compatibility, the same handler is bound to `/actions/workflows/{workflow_file}/badge.svg`. (Both under `/{username}/{repo}/`, of course)

CanisHelix commented:
As of approximately 2 weeks ago shields.io supports release/language badges for Gitea/Forgejo out of the box. More endpoints will be added soon, most likely next will be issues/stars once I get to it. Not sure if this information helps this PR or not but just wanted to share this.

algernon commented:
> As of approximately 2 weeks ago shields.io supports release/language badges for Gitea/Forgejo out of the box. More endpoints will be added soon, most likely next will be issues/stars once I get to it. Not sure if this information helps this PR or not but just wanted to share this.

Sweet, that means that a lot of this PR will be unnecessary soon. I suppose only the Actions parts should remain then, because there is no API where the information could be collected from by shields.io to support them natively.

In the longer run, perhaps the better action would be to add those APIs, and have shields.io support Build (Action) badges for Gitea/Forgejo out of the box, too.
